[Operator Versioning][Edge] Codegen upgrader_mobile.cpp

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69194

From operator version map and upgrader torchscript, generate upgrader_mobile.cpp file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

Commit:
[Operator Versioning][Edge] Codegen cpp file with yaml file



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen cpp file with yaml file"


Given a yaml file input with bytecode, generate upgrader_mobile.h header file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

iseeyuan reviewed (commented):
I think we only need to commit one generated file, either yaml or the header file. Looks like committing the header file makes more sense?

cccclai commented:
> I think we only need to commit one generated file, either yaml or the header file. Looks like committing the header file makes more sense?

Currently no files are checked in. The checked in two files are for testing purposes at the moment. Likely Tugsuu will need to use this file to generate cpp.

iseeyuan reviewed (approved):
LGTM.

Commit:
Update on "[Operator Versioning][Edge] Codegen cpp file with yaml file"


Given a yaml file input with bytecode, generate upgrader_mobile.h header file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

gmagogsfm commented on tools/codegen/operator_versions/gen_mobile_upgraders.py:
```diff
@@ -0,0 +1,285 @@
+
+
+def main(argv):
+    parser = argparse.ArgumentParser(description="Generate used operators YAML")
```
Description seems to be for a different argument?

cccclai replied:
Oh good call. Update in the latest commit.

gmagogsfm commented on tools/codegen/operator_versions/gen_mobile_upgraders.py:
```diff
@@ -0,0 +1,285 @@
+  int index;
+};
+
+// From operator_versions.yaml
+const std::unordered_map<std::string, std::vector<Upgrader>> kOperatorVersionMap(
+    {
+        {std::string("aten::div.Tensor"),
+         std::vector<Upgrader>({Upgrader({0, 3, "div_Tensor_0_3", 0})})},
+        {std::string("aten::div.Scalar"),
+         std::vector<Upgrader>({Upgrader({0, 3, "div_Scalar_0_3", 1})})},
+        {std::string("aten::div.out"),
+         std::vector<Upgrader>({Upgrader({0, 3, "div_out_0_3", 2})})},
+        {std::string("aten::div_.Tensor"),
+         std::vector<Upgrader>({Upgrader({0, 3, "div__Tensor_0_3", 3})})},
+        {std::string("aten::div_.Scalar"),
+         std::vector<Upgrader>({Upgrader({0, 3, "div__Scalar_0_3", 4})})},
+    });
```
Why are upgrader definitions hard coded here?
@tugsbayasgalan I think you mentioned that operator_versions.yaml will be deleted, then what will replace it? It seems we are still referencing it here.

cccclai replied:
Yeah this python will take an input bytecode yaml and output cpp. 

In the latest commit, `kOperatorVersionMap` is generated from `torch._C._get_operator_version_map()`, which is populated from `caffe2/torch/csrc/jit/operator_upgraders/version_map.h`  in https://github.com/pytorch/pytorch/pull/68772.

cccclai replied:
In https://github.com/pytorch/pytorch/pull/68005, `def generate_bytecode(file_name)` will generate a bytecode yaml file, I will append this function at the end of `generate_bytecode` to generate a bytecode cpp instead.

cccclai replied:
The reason why codegen another version map is that it needs to map the the index in `std::vector<ByteCodeFunctionWithOperator> kUpgraderByteCode`. When replacing the operator in bytecode
`OP`, `X`, `N`, we can replace `OP` with `CALL` and `X` with the according index in `kUpgraderByteCode`. 

Maybe there is a more convenient way to replace. What I describe above is the implementation in https://github.com/pytorch/pytorch/pull/67731

Commit:
Update on "[Operator Versioning][Edge] Codegen cpp file with yaml file"


Given a yaml file input with bytecode, generate upgrader_mobile.h header file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen cpp file with yaml file"


Given a yaml file input with bytecode, generate upgrader_mobile.h header file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

tugsbayasgalan commented on test/mobile/test_upgrader_bytecode_table_example.h:
```diff
@@ -0,0 +1,329 @@
+
+// From operator_versions.yaml
+
+const std::unordered_map<std::string, std::vector<Upgrader>> kOperatorVersionMap(
```
Can this be a different name if we are using kOperatorVersionMap for the actual version_map.h

cccclai replied:
Yeah I will rename it

tugsbayasgalan commented on tools/codegen/operator_versions/gen_mobile_upgraders.py:
```diff
@@ -0,0 +1,357 @@
+        for upgrader_entry in version_map[op_name]:
+            # Split a string by "_" and filter empty string in the list
+            # For example: "div__Scalar_0_3" => ['div', 'Scalar', '0', '3']
+            upgrader_info = list(filter(lambda token: token != "", upgrader_entry.upgrader_name.split('_')))
```
I think it is not safe to rely on the upgrader name to figure out the start and end. I think it rather should be based on version bump value. So for example, if the version bump happened at 4, we know that the range will be from 0 to 3.

Commit:
Update on "[Operator Versioning][Edge] Codegen cpp file with yaml file"


Given a yaml file input with bytecode, generate upgrader_mobile.h header file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen cpp file with yaml file"


Given a yaml file input with bytecode, generate upgrader_mobile.h header file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen cpp file with yaml file"


Given a yaml file input with bytecode, generate upgrader_mobile.h header file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen cpp file with yaml file"


Given a yaml file input with bytecode, generate upgrader_mobile.h header file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen cpp file with yaml file"


Given a yaml file input with bytecode, generate upgrader_mobile.h header file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen cpp file with yaml file"


Given a yaml file input with bytecode, generate upgrader_mobile.h header file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen cpp file with yaml file"


Given a yaml file input with bytecode, generate upgrader_mobile.h header file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen upgrader_mobile.cpp"


From operator version map and upgrader torchscript, generate upgrader_mobile.cpp file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen upgrader_mobile.cpp"


From operator version map and upgrader torchscript, generate upgrader_mobile.cpp file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen upgrader_mobile.cpp"


From operator version map and upgrader torchscript, generate upgrader_mobile.cpp file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen upgrader_mobile.cpp"


From operator version map and upgrader torchscript, generate upgrader_mobile.cpp file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen upgrader_mobile.cpp"


From operator version map and upgrader torchscript, generate upgrader_mobile.cpp file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Codegen upgrader_mobile.cpp"


From operator version map and upgrader torchscript, generate upgrader_mobile.cpp file. It also includes a unit test.



Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

[ghstack-poisoned]
